### PR TITLE
Add repl short url support for #66

### PR DESCRIPTION
**Pull request by brigand**
Edit: parse.com is shutting down within the year, so this PR doesn't make sense anymore.

**WIP**
 - [x] functionality 
 - [x] styling (useable, I'm not a designer)
 - [ ] update parse client id/key

This adds the ability to use 'short' urls with the repl.  Really they're not very short, but they are unique, persistent, and the length isn't dependant on the input.  The implementation uses Parse.com, but it could be swapped out with gists or similar in the future.

Example URL:

```
/repl/#?short=dOWCPHsLgJ
```

I'm opening this before it's done so @sebmck can create an application in parse.com and give the client id/key.  Once one is saved with that id, you need to go into the data viewer and change the Class Level Permissions for ReplItem... they're very very permissive by default (like having direct database access on the client, permissive).

e.g. if you named it 'babel-repl': https://parse.com/apps/babel-repl/collections#class/ReplItem

Class Level Permissions for ReplItem in parse.com UI

And then in settings on the top nav bar, switch it into production mode and disable client side class creation.

parse.com application settings screen

After that, it'll just work and we'll never hit the free limit.

#66

### CODE REVIEW

File: scripts/repl.js [hunk 1]
```diff
@@ -57,6 +62,86 @@
   };
 
   /*
+   * Utils for working with parse.com for persistence
+   */
+  function ParseUtils () {}
+
+  var __parseLoadingCallbacks = null;
```
Comment (thejameskyle) [inline scripts/repl.js]: Is there a reason this can't be added as an instance property of ParseUtils?

File: scripts/repl.js [hunk 2]
```diff
@@ -57,6 +62,86 @@
...
+    this.afterLoad(function(ReplItem){
+      var query = new Parse.Query(ReplItem);
+      query.get(id)
+        .then(function (data) {
+        // give us a plain object without anything parse specific, making
+        // .get's output the same as .save's input
+        var cleanData = data.toJSON();
+        delete cleanData.objectId;
+        delete cleanData.createdAt;
+        delete cleanData.updatedAt;
```
Comment (thejameskyle) [inline scripts/repl.js]: It's probably fine just to let this fall through.

File: scripts/repl.js [hunk 3]
```diff
@@ -284,7 +371,34 @@
   /*
    * Initialize the REPL
    */
-  var repl = new REPL();
+  var initialQuery = UriUtils.parseQuery();
+  if (initialQuery && initialQuery.short) {
```
Comment (thejameskyle) [inline scripts/repl.js]: @sebmck Do you think it'd make sense to require all of these to be shortened urls?

File: scripts/repl.js [hunk 4]
```diff
@@ -301,10 +415,33 @@
...
+        alert('Could not shorten url, see developer console console.');
+        return;
+      }
+
+      UriUtils.updateQuery({short: id});
+
+      repl.$shortenUrlButton.css({color: '#70ff70'});
+      setTimeout(function (){
+        repl.$shortenUrlButton.css({color: ''});
+      }, 1500);
```
Comment (thejameskyle) [inline scripts/repl.js]: What exactly is this timeout for?

Comment (brigand) [inline scripts/repl.js]: Just the css transition.

File: scripts/repl.js (hunk 3 above)
Comment (brigand) [inline scripts/repl.js]: It doesn't require the url to be shortened, it's just handling the case where it is.  Otherwise it falls through to the existing behavior.

File: scripts/repl.js (hunk 2 above)
Comment (brigand) [inline scripts/repl.js]: I believe it'll end up in state, which will end up in the URL.  I'll confirm it, though.

File: scripts/repl.js (hunk 1 above)
Comment (brigand) [inline scripts/repl.js]: Not currently, but it'd be fragile.

```js
new ParseUtils().load().afterLoad(f);
new ParseUtils().load().afterLoad(g);
```

This loads parse once, both f and g are called when it's loaded.  Changing to to an instance property would load parse twice and maybe break things.  It could be made a property of ParseUtils if that's prefered.

**Comment by brigand** [comment]
Thanks for the comments.  I missed the email notification, so sorry for the late response.

File: scripts/repl.js (hunk 2 above)
Comment (brigand) [inline scripts/repl.js]: Yeah, I ended up with: http://localhost:8080/repl/#?code=5%2B5%2B1&evaluate=true&experimental=true&loose=false&playground=false&spec=false&objectId=6kDKy2aEul&createdAt=2015-03-28T07%3A35%3A23.598Z&updatedAt=2015-03-28T07%3A35%3A23.598Z

File: scripts/repl.js (hunk 3 above)
Comment (thejameskyle) [inline scripts/repl.js]: I meant instead of the old query params (`?experimental=true&code=foo`)

File: scripts/repl.js (hunk 1 above)
Comment (thejameskyle) [inline scripts/repl.js]: Why not create a factory function that returns a singleton? Is there even a case where we'd have to `ParseUtils`?

File: scripts/repl.js (hunk 2 above)
Comment (thejameskyle) [inline scripts/repl.js]: Alright np

File: scripts/repl.js (hunk 4 above)
Comment (thejameskyle) [inline scripts/repl.js]: cool

File: scripts/repl.js (hunk 1 above)
Comment (brigand) [inline scripts/repl.js]: It's always constructed here:

```js
function REPL () {
  this.shortener = new ParseUtils();
```

And conditionally here:
```js
if (initialQuery && initialQuery.short) {
  new ParseUtils().get(initialQuery.short, function (err, data) {
```

It could be done in an iife to create a singleton if that's preferred.  I was just trying to stick to the existing style as much as possible.

File: scripts/repl.js (hunk 3 above)
Comment (brigand) [inline scripts/repl.js]: Meaning it should leave the url at /repl/ until they hit save, and only support linking with short urls?

Comment (thejameskyle) [inline scripts/repl.js]: Yes, I'm leaning towards that being the right thing to do, but I also don't want to break existing URLs.

Comment (brigand) [inline scripts/repl.js]: Okay, I'll try to do this later today.  It can load the old urls, and just have the url always be `#` or `#?short=...`.

Comment (thejameskyle) [inline scripts/repl.js]: Cool! Keep it as a separate commit please. :smile:

**Comment by brigand** [comment]
parse.com is shutting down within the year, so this PR doesn't make sense anymore.
